Add complex support for `torch.{acosh, asinh, atanh}`

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50387 Add complex support for `torch.{acosh, asinh, atanh}`**

Differential Revision: [D25947496](https://our.internmc.facebook.com/intern/diff/D25947496)

anjali411 committed:
Add complex support for

[ghstack-poisoned]

anjali411 committed:
Update on "Add complex support for `torch.{cosh, sinh, tanh}`"




[ghstack-poisoned]

mruberry commented on test/test_torch.py:
```diff
@@ -6904,9 +6904,10 @@ def inner(self, device, dtype):
-    ('acosh', '', lambda t, d: _small_3d(t, d) + 1, lambda t, d: [], 1e-3, 1e-2, 1e-5, torch.testing.get_all_fp_dtypes()),
-    ('asinh', '', _small_3d, lambda t, d: [], 1e-3, 1e-2, 1e-5, torch.testing.get_all_fp_dtypes()),
-    ('atanh', '', _small_3d, lambda t, d: [], 1e-3, 1e-2, 1e-5, torch.testing.get_all_fp_dtypes()),
+    ('acosh', '', lambda t, d: _small_3d(t, d) + 1, lambda t, d: [], 1e-3, 1e-2, 1e-5,
```
These were recently removed (they're redundant with the OpInfo tests). You can just revert changes to test_torch.py for this PR.

mruberry commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -336,23 +336,23 @@ static void cosh_kernel(TensorIterator& iter) {
 }
 
 static void acosh_kernel(TensorIterator& iter) {
```
CUDA dispatches also need an update.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -807,9 +807,9 @@ def sample_inputs_fliplr_flipud(op_info, device, dtype, requires_grad):
-                   dtypes=all_types_and(torch.bool),
-                   dtypesIfCPU=all_types_and(torch.bool),
-                   dtypesIfCUDA=all_types_and(torch.bool, torch.half, torch.bfloat16),
+                   dtypes=all_types_and_complex_and(torch.bool),
```
Like other trig operators these will probably need to skip the reference numerics test for complex on CPU. In the future we should revise that test to have an easier version, like test_small_reference_numerics, so we only have to skip some values in testing.

anjali411 committed:
Update on "Add complex support for `torch.{acosh, asinh, atanh}`"




[ghstack-poisoned]

anjali411 committed:
Update on "Add complex support for `torch.{acosh, asinh, atanh}`"




[ghstack-poisoned]

mruberry commented on test/test_torch.py:
```diff
@@ -6906,13 +6906,6 @@ def inner(self, device, dtype):
         torch.testing.get_all_dtypes(include_complex=False, include_bool=False), [torch.bfloat16]),
     ('sign', '', _small_3d, lambda t, d: []),
     ('logit', '', _small_3d, lambda t, d: [], 1e-3, 1e-2, 1e-5, torch.testing.get_all_fp_dtypes()),
-    ('sqrt', '', _small_3d, lambda t, d: [], 1e-3, 1e-2, 1e-5, torch.testing.get_all_fp_dtypes(), [torch.bfloat16]),
```
Thank you for cleaning these up.

mruberry commented:
Looks like this will have to skip test_fn_grad_acosh_cuda_complex128 - TestGradientsCUDA on pytorch_windows_vs2019_py36_cuda10.1_test2, too.

I'm not totally sure what's going on there.

mruberry reviewed (approved):
Cool! Looks like this just needs one more skip.

For that skip would you file an issue and copy the issue's URL into a comment over the skip?

anjali411 committed:
Update on "Add complex support for `torch.{acosh, asinh, atanh}`"




[ghstack-poisoned]

anjali411 committed:
Update on "Add complex support for `torch.{acosh, asinh, atanh}`"


Differential Revision: [D25947496](https://our.internmc.facebook.com/intern/diff/D25947496)

[ghstack-poisoned]
